realtek: clock support for RTL930X (v4)

plappermaul opened the pull request:
a clock driver to rule them all ...

yarrick commented on target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig:
```diff
@@ -14,6 +14,7 @@ config COMMON_CLK_RTL83XX
 	  This includes chips in the RTL838x series, such as RTL8380, RTL8381,
 	  RTL832, as well as chips from the RTL839x series, such as RTL8390,
-	  RT8391, RTL8392, RTL8393 and RTL8396.
+	  RT8391, RTL8392, RTL8393 and RTL8396 and chips of the RTL839X series
```
chips of the RTL930X series?

plappermaul replied:
fixed

plappermaul commented:
Attached v2 of the PR with following changes

- unlocking of RTL838X registers
- Moved documentation to a new yaml file

lucize commented:
boot tested with RTL9302B, not tested extensively

svanheule reviewed (changes requested):
The commit "realtek: clock driver enhancement for RTL930X" does so many things at once, it's really not reviewable, nor are the changes useful to bisect later. Please split this into smaller patches that fix or add one issue.

Commit "realtek: switch RTL930X DT to new clock driver" then introduces new DT properties, which haven't been announced in any commit message before. I assume they are hidden in between the changes in the first patch.

Commit "realtek: fixes in RTL83XX DT" should come before the new features. The RTL83xx fixes should probably even be a separate PR, since the new features (RTL930x support) are less urgent IMHO.

Commit "realtek: DT documentation for clock driver" should remove the comment from the source file, and translate that into a DT binding in one commit. Now you remove the documentation in the first commit, but it's only added back at the very end of the series.

plappermaul committed:
realtek: clock driver: rename variables

Use our common prefix for several variables

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: simplify magic register values

For simplicity we use register magic values. We simply read/write full registers
but only some of the bits are relevant. So we might overwrite data that is
different depending on the vendor. For simplification we will handle constants
as follows:

- constants are right shifted and contain only relevant bits
- the code will left shift the values as needed
- the code will run a read-modify-write cycle
- As syncs are not needed for registers we can drop that part

Btw. SRAM can be access cached/uncached. When using parts of it make clear that
we access the cached part.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: get away with register macros

Just use plain io functions. In a future version we will simplify the
(void *) modifications but for now keep the commit simple.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: care about locked register

some RTL838X devices are shipped with locked registers. Remove the locks
so we can read/write the required PLL registers.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: reorganize frequency reading

To better adapt to future devices reorganize the register reading. A small
macro helps us to build a switch statement for SOC and clock.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: provide crystal/oscillator clock

Until now the driver depends on an external DT oscillator definition.
For now we can not see any benefit of setting that in the device tree.
Lots of other Linux kernel drivers hardcode these values. Usually the
name xtal (crystal) is used there. Align that and register a fixed
frequency clock (named xtal_clk) in the driver.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: reorganize documentation

Move the documentation into the device tree.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: cut comments to 80 chars

Try to keep line size with kernel recommendation.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: cosmetic fixes

Especially do the following:

- Fix wrong DRAM function definitions (although never used).
- Print speed ranges during startup only if changing is really supported.
- Add some namespace prefixes.
- Use kernel common type "unsigned long" for clock rate

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: allow multiple clock parents

RTL930X LXB clock is not derived from the crystal. In a future commit
we will add that and need the possibility to derive a clock from something
else than the crystal. Make this configurable per SOC type.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: add switch clock

The internal switch clock will become relevant when we add RTL930X support.
Add it to the list.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: adapt RTL838X DT

The clock driver is self contained now. Remove old oscillator definition
from the DT. We need a small SRAM scratchpad area for future use. Add it.
While we are here adapt the SRAM size to the real 128K as the manuals say.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: adapt RTL839X DT

The clock driver is self contained now.	Do the same as for RTL838X.
Remove the oscillator and add some scratch SRAM area for future use.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: enhancement for RTL930X

The previous version only covered the RTL83XX series. Add all required
features to support also the RTL930X devices.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: enable basic RTL930X config for cpufreq

Before we enable the driver for the new target we must tell the config that
RTL930X allows CPU frequency changing.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: add for RTL930X target

Make use the new clock driver for RTL930X target devices. Add its primary
consumer (cpufreq-dt) too. To be careful just set the default governor to
userspace. As we rely on SRAM activate that module too.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: clock driver: switch RTL930X DT syntax

Use new DT clockdriver syntax for RTL930X target. To make it work we
need to change some nodes:

- define SRAM
- add clock controller
- Add second CPU node for future SMP use
- map all devices to new clocks
- Remove dummy LXB clock
- add CPU OPP table

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
@svanheule Dont want to complain but dissecting and recompiling each and every patch for the updated PR with this slow build system just relocates the work.

plappermaul commented:
Any complains left after the split?

svanheule commented:
Sorry, haven't had the time to look at this properly again last week. I'll try to provide some comments this week.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl839x-sram.S:
```diff
@@ -16,16 +16,17 @@
- * Functions start here and should avoid access to normal memory. REMARK! Do not forget about
- * stack pointer and dirty caches that might interfere.
+ * Functions start here and should avoid access to normal memory. REMARK! Do not
+ * forget about stack pointer and dirty caches that might interfere.
```
Spurious change, not really relevant for this patch.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -600,8 +592,12 @@ static struct clk_hw *rtcl_get_clkhw(struct of_phandle_args *clkspec, void *prv)
 
+	clk = clk_register_fixed_rate(NULL, "xtal_clk", NULL, 0, RTCL_XTAL_RATE);
+	clk_register_clkdev(clk, "xtal_clk", NULL);
+
```
Aside from a simpler devicetree definition, what's the benefit of making the driver provide the fixed rate clock input?

In case of a dual-SoC system and synchronous ethernet support (which these chips do appear to support), it might be possible that the secondary SoC is not driven by an external crystal, but could be driven by some external clock generator that needs to be set up first. Having that in the devicetree would provide a proper clock hierarchy. Whereas hard-coding the clock input frequency here omits that explicit relation between the two.

Unless I am misunderstanding the point of this patch, I don't think omitting the 25MHz clock from the board definition (i.e. DTS) is the "right" thing to do. The clock/crystal is an external component, so it should be defined as such IMHO.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -264,13 +259,13 @@ struct rtcl_round_set {
-		RTCL_ROUND_SET(50000000, 200000000, 50000000)
+		RTCL_ROUND_SET(400000000,	850000000,	25000000),
+		RTCL_ROUND_SET(100000000,	400000000,	25000000),
+		RTCL_ROUND_SET(50000000,	200000000,	50000000)
```
These cosmetic changes aren't really related to making the 25MHz clock internal to the driver.

svanheule commented on target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml:
```diff
@@ -0,0 +1,76 @@
+description: |
+  The driver provides basic clock support for the central core clock unit (CCU)
+  and its PLLs inside the RTL838X, RTL8389X and RTL930X SOC. All provided clocks
+  are derived from an oscillator that runs at a fixed speed of 25 MHz. Currently
```
oscillator or external clock

plappermaul replied:
datasheet states: "25MHz Crystal Clock"

svanheule replied:
![image](https://user-images.githubusercontent.com/6670918/189995432-37c73bf5-3033-44b9-8141-70e1b4aa287d.png)
"When using a crystal", so might also be an external clock driver.

svanheule commented on target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml:
```diff
@@ -0,0 +1,76 @@
+          opp-hz = /bits/ 64 <425000000>;
+      };
+      ...
+  }
```
You should make this a second example:
- One simple example, with a minimal DT setup
- One extended example, with all possible options added.

That way, the DT tooling will be able to detect errors in your examples too.

Please also try to verify that your binding is valid: https://www.kernel.org/doc/html/latest/devicetree/bindings/writing-schema.html (I can't spot any obvious errors)

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -34,15 +34,13 @@
-#define REG_CTRL0		0
-#define REG_CTRL1		1
-#define REG_COUNT		2
+#define RTCL_REGCNT		2
```
Can you put this in the other patch with prefix changes?

And REG_CTRL{0,1} weren't used then?

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -74,7 +72,7 @@ static const int rtcl_regs[RTCL_SOCCNT][REG_COUNT][CLK_COUNT] = {
 
 struct rtcl_reg_set {
-	unsigned int rate;
+	unsigned long rate;
```
Why the change? Isn't (unsigned) int the same size as (unsigned) long anyway?

plappermaul replied:
wherever rate is used we take the long type. So be consistent here.

svanheule replied:
OK, it appears you changing for consistency with the clock framework, which uses "unsigned long" for clock rates. Please try to put these "why"s in the commit message, so we don't have to spend time wondering why you're making changes like this.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -1,4 +1,4 @@
-// SPDX-License-Identifier: GPL-2.0-only
+/* SPDX-License-Identifier: GPL-2.0-only */
```
SPDX identifiers with `//` are correct in a source file.
See: https://www.kernel.org/doc/html/v4.18/process/license-rules.html#license-identifier-syntax

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -238,9 +239,9 @@ struct rtcl_clk {
-	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "LXB")
+	RTCL_CLK_INFO(CLK_CPU, "cpu_clk", "xtal_clk", "xtal_clk", "CPU"),
+	RTCL_CLK_INFO(CLK_MEM, "mem_clk", "xtal_clk", "xtal_clk", "MEM"),
+	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "xtal_clk", "LXB")
```
Why have two identical names?

This patch doesn't seem to line up with the commit message either... If a clock may not necessarily have a fixed clock as source, why not just use the framework by putting everything in the DTS?

svanheule commented on target/linux/realtek/files-5.10/include/dt-bindings/clock/rtl83xx-clk.h:
```diff
@@ -7,9 +7,10 @@
+#define CLK_CPU		1
+#define CLK_MEM		2
+#define CLK_LXB		3
+#define CLK_COUNT	4
```
Why are you changing the values of these `#define`s? Adding `#define CLK_SWITCH 4` would be a smaller change, and the values are arbitrary anyway.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -14,18 +14,29 @@
 #define RTL838X_PLL_GLB_CTRL			(0x0fc0)
 #define RTL838X_PLL_CPU_CTRL0			(0x0fc4)
 #define RTL838X_PLL_CPU_CTRL1			(0x0fc8)
+#define RTL838X_PLL_CPU_MISC_CTRL		(0x0fcc)
```
I don't see the `*_MISC_CTRL` registers being used here. I suppose it doesn't hurt to have the defined, but they are sort of sneaked in with another change.

svanheule commented on target/linux/realtek/dts-5.10/rtl838x.dtsi:
```diff
@@ -62,17 +62,9 @@
 		compatible = "realtek,rtl8380-clock";
 		#clock-cells = <1>;
-		clocks = <&osc>;
-		clock-names = "ref_clk";
```
I don't think being able to omit this, is worth the extra boilerplate in the clock driver.

plappermaul replied:
This is only needed if oscillator stays in DT as fixed clock

svanheule commented on target/linux/realtek/dts-5.10/rtl838x.dtsi:
```diff
@@ -262,10 +254,16 @@
+			compatible = "realtek,scratch";
+			reg = <0x0 0x20>;
+			pool;
+		};
```
Any reason this has/needs a specific compatible?

svanheule replied:
OK, this is something that would go into Documentation/devicetree/bindings/sram/sram.yaml then. Maybe add a label too.; something that indicates this is used by the clock driver to update the PLL registers: e.g.`label = "clock-config-update";` 

Can you use `protect-exec;` or is that not possible when also writing the blob from the driver?

svanheule replied:
Unless I'm missing something, these new sram-scratch nodes also aren't used anywhere. Why are you adding them? If you need them for an upcoming feature, these should be included with those patches.

plappermaul replied:
They are used in the 930x assembler code (SRAM_BASE_UNCACHED) to "reset" SRAM communication. References in other DTs have been removed.

svanheule commented on target/linux/realtek/dts-5.10/rtl838x.dtsi:
```diff
@@ -262,10 +254,16 @@
 		#address-cells = <1>;
 		#size-cells = <1>;
-		ranges = <0 0x9f000000 0x10000>;
+		ranges = <0 0x9f000000 0x20000>;
```
Separate commit with Fixes: tag? Then that can be merged ahead of some of the bigger changes.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -247,12 +314,16 @@ struct rtcl_clk {
+	RTCL_CLK_INFO(CLK_SW, "sw_clk", "xtal_clk", "xtal_clk", "xtal_clk", "SW"),
+	RTCL_CLK_INFO(CLK_CPU, "cpu_clk", "xtal_clk", "xtal_clk", "xtal_clk", "CPU"),
+	RTCL_CLK_INFO(CLK_MEM, "mem_clk", "xtal_clk", "xtal_clk", "xtal_clk", "MEM"),
+	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "xtal_clk", "sw_clk", "LXB")
```
Aha, this is why you have all those duplicated parent clocks. You should be able to encode this in the DTS.

plappermaul replied:
Dependency SW/LXB cannot be encoded in the DT as both are provided by the driver itself.

svanheule replied:
I think you should be able to self-reference the clock node:
```
ccu: clock-controller {
	compatible = "realtek,rtl8380-clock";
	#clock-cells = <1>;
	clocks = <&osc>, <&ccu CLK_SW>;
	clock-names = "ref_clk", "sw_clk";
};

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:
```diff
@@ -79,12 +94,25 @@
 
 #define RTL_SRAM_MARKER				(0x5eaf00d5)
 #define RTL_SRAM_BASE_CACHED			(0x9f000000)
+#define RTL_SRAM_BASE_UNCACHED			(0xbf000000)
```
You've removed the `sync` ops earlier, shouldn't you use uncached memory there as well then?

plappermaul replied:
sync is a no longer needed leftover from c compiled sources that I copy/pasted to assembler.

svanheule replied:
Why are they no longer needed though? I don't recall reading that in the commit message.

svanheule reviewed (changes requested):
Please avoid abbreviations like "reorg" in you commit titles/messages. When you're going over the commit message, also have a look at capitalising the sentences, and adding full stops at the end.

plappermaul commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -484,6 +481,26 @@ static const struct clk_ops rtcl_clk_ops = {
+	if ((reg & enable) != enable) {
+		reg |= enable;
+		iowrite32(reg, (void *)RTL838X_INT_RW_CTRL);
+		pr_warn("%s: registers unlocked\n", __func__);
```
Why should I warn about unlocked registers? That is the desired state and if set correctly by someone else (vendor) everything is fine. Opening up the registers should have a meaningful level.

svanheule replied:
That's exactly my point. "unlocked registers" is the desired state, so that's what you set. "locked registers" is just the default state, so not necessarily an indication that something is wrong, which would warrant a `pr_warn()`. I don't think you need to inform the user at all, just unlock the registers and be done with it.

plappermaul commented:
Hi Sander,

maybe not the perfect place, but best I can see. While I appreciate your reviews and feedback I get the impression that developing new stuff for devices and pushing it to OpenWrt is not the favourite way to go. Why? Most of the feedback I got until now was of formal nature. This is not a problem as long as it does not get into a chair cycle discussion for that spare time is too valuable. Examples?

- A fixed frequency clock inside the driver is not favoured. Mainline kernel has >100 of these.
- In v1 a DT documentation is encouraged, v2 provides it (not much I can see elsewhere in OpenWrt) and still this is not good enough.
- changing variable types/names should be explained in commit messages
- a 7 line commit changing a timeout loop results in a discussion about capitalisation an where/how to reference commits
- The CAMEOTAG patch came close to a mathematical lecture

The first driver version went over 9 iterations and more than 200 comments until it got accepted. No problem for the initial round. But looking at the current feedback for 17 small patches for v2 we might top the numbers by far. Just for cleaning things up and adding support for a new device inside the driver?

To my personal pipeline (irq driver fix, 930x SMP activation, the new timer driver, 5.15 testing) I added the first stuff for implementing a hw accellerated AES/SHA1 crypto driver for the engine inside the 838x and 930x. With hours invested in understing asynchronous hash/cipher handling I'm thinking about the discussions and must-have-it-that-way-requests that will start afterwards.

To quote @robimarko from the old thread "the maintainers opinion pretty much equates to the project's opinion". I agree with that 100%. With that in mind the 5.15 RFC discussion "don't want to merge with L2 learning disable either, ... <he> wasn't able to figure out yet where things go wrong, but hopefully someone else can." feels strange. 

Writing this 40 days after Birger left seems hard but a good point for thinking how to go on from here. Development by proxy feels odd. And so it does whenever someone is complaining about old bad code and in the same sentence is setting highest expectations for new changes. I prefer reasonable patches that might not be perfect but keep it rolling.

svanheule commented:
Hi Markus,

As much as I would like to do development myself, I currently do not have the time do so, since this is a spare time thing for me. That means "development by proxy" is what I am currently relegated to. I realize that my comment on the 5.15 issue can come across as "you guys do all the hard work, I'll just sit back and watch". That was not what I meant to say, so I must apologize. Hiroshi had these changes set aside for a while, and we would need to make the transition at some point. I figured a pull request may get it the extra sets of eyes it needed to pass the final hurdles.

Big changes like this patch set are hard to review, because they do so many things at once. I personally find it discouraging to have to pour through change dumps, and we *have* needed to hunt down bugs introduced by big, badly explained change sets before. So I really appreciate that you took the time and effort to split the changes into smaller patches. Now I think I have a better idea of the changes you want to make. Most changes are just fine, by the way, but I commented with most things that came to mind, both major and minor, so we don't end up at v10 before I start with the formalities.

Even if things aren't perfectly to your liking the first time around, I think it can sometimes be preferable to keep them that way if possible, to not distract from the actual changes you need to make to go forward. For example, the devicetree binding is a hard requirement upstream, but I am of the opinion that here, for OpenWrt, a comment in the source is sufficient. Yes, a binding will be required, but I am happy to keep things as they are here. My apologies for not making that sufficiently clear. If you want to include a proper binding with the other changes now, I won't object to that, but it also means I will take (and need) time to look at them and give feedback.

I am happy to see more people, like you, providing new code. As much as I can, I try to understand the code you are providing, and new features they expose. As you noticed, that means certain choices you make may pass the first time, because it's impossible for me to understand every little detail of a completely new driver, in a subsystem I've never used before (which means the majority of subsystems). When you make changes afterwards, I am left wondering why things need to change if I don't see the "why" of a change in the commit message.

What I *try* to do as a maintainer (but certainly do not always succeed at), is to keep things consistent among different developers. You have your preferences, and so does Olliver, or anybody else contributing to the project. In my opinion, consistency is key to also enable others to contribute back to the project. So yes, I fear that will inevitably include things that may come across as mere formalities and nit-picking. While I agree that this can be frustrating, once code is in OpenWrt (or the upstream kernel for example), it becomes part of the project. I'm afraid we can't just have the project be somebody's personal playground, just to give some code more exposure.

Please don't get discouraged by the process. I do appreciate your contributions to the project, and hope you decide to keep contributing, even if things are slow sometimes.

plappermaul commented:
To get forward here I'll take your word for it "Most changes are just fine". So I have come to the following lonely descision.

- I changed the commit messages to better reflect their content.
- I did not (and will not) change the content. I do not like to rearrange 17 commits just to come to the same result.
- After your better understanding of the code, I ask you to have a quick look at the complete final package (at least C file).
- If something really should get rework (and is worth the effort), drop a note here or in the chat for me.
- I will fix it in a follow-up

This will keep maintainer & developer overhead small. So feel free to merge.

svanheule commented on target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:
```diff
@@ -484,6 +481,26 @@ static const struct clk_ops rtcl_clk_ops = {
+ * benefit of always unlocking/locking these registers just open up everything.
+ */
+	enable = RTL838X_INT_RW_CTRL_READ_EN | RTL838X_INT_RW_CTRL_WRITE_EN;
+	reg = ioread32((void *)RTL838X_INT_RW_CTRL);
```
This crashes my device on boot, since it's missing the switch core address space offset.

plappermaul replied:
fixed, thanks for testing

svanheule commented on target/linux/realtek/dts-5.10/rtl930x.dtsi:
```diff
@@ -8,14 +10,57 @@
+			operating-points-v2 = <&cpu_opp_table>;
+		};
+
+		cpu@1 {
```
Same comment as for the sram-scratch nodes. If this is still unused at this point, please leave it out of the change set. AFAICT, adding the second VPE is not required for switching over to the new clock driver; which this commit is supposedly doing.

plappermaul replied:
removed

plappermaul commented:
updated as requested.

plappermaul commented:
@svanheule maintainer ping after 11 days.

plappermaul commented:
@svanheule Another 11 days over without a single bit of feedback. A simple "will not merge because of ..." will be enough.

robimarko commented:
I am sure he is just really busy, OpenWrt rarely pays bills for devs

plappermaul commented:
rebased

plappermaul commented:
> I am sure he is just really busy, OpenWrt rarely pays bills for devs

Same for me.
